src/odemis/driver/tfsbc.py

```diff
@@ -0,0 +1,130 @@
+BEAMDEFL_LY = 1  # lower y beam deflection element control
+BEAMDEFL_UX = 2  # upper x beam deflection element control
+BEAMDEFL_UY = 3  # upper y beam deflection element control
+
+RANGE_AMPS = [-42.2e3, 42.2e3]
+
+class BeamController(model.HwComponent):
+    """
+    Driver for the Thermofischer beam controller.
+    This class provides the .shift VA containing a tuple of two floats which describe
+    the x and y beam offset in m.
+    """
+
+    def __init__(self, name, role, amps_per_m, zero_offset_amps,
+                 rs485_serial_num, **kwargs):
+        """
+        :param port (str): port (e.g. /dev/ttyUSB0)
+        :param amps_per_m (float): conversion factor amplitude/meter
+        :param zero_offset_amps (float): amplitude for which beam deflection is 0
+        :param rs485_serial_num (str): serial number of RS485 adapter
+        """
+        # .hwVersion, .swVersion not available
+        model.HwComponent.__init__(self, name, role, **kwargs)
+
+        self._amps_per_m = amps_per_m
+        self._zero_offset_amps = zero_offset_amps
+
+        # Find port by RS485 adapter serial number
+        for p in serial.tools.list_ports.comports():
+            if rs485_serial_num in p.hwid:
+                # Found it!
+                port = p
+                break
+        else:
+            raise IOError("RS484 adapter with serial number %s not found!" % rs485_serial_num)
+
+        # Modbus connection
+        self._modbus = ModbusSerialClient(method='rtu', port=port,
+                                          baudrate=BAUDRATE, timeout=TIMEOUT,
+                                          stopbits=STOPBITS, parity=PARITY,
+                                          bytesize=BYTESIZE)
+
+        # Calculate range
+        range_amps = numpy.subtract(RANGE_AMPS, zero_offset_amps)
+        range_m = numpy.divide(range_amps, amps_per_m)
+
+        # Default value: zero offset (in m)
+        default_shift = tuple(numpy.divide(zero_offset_amps, amps_per_m))
+        self.beamShift = model.TupleContinuous(default_shift, range_m,
+                                           cls=(int, float), unit="m",
+                                           setter=self._setShift)
+
+    def _setShift(self, value):
+        """
+        :param value (float, float): x, y shift from the center (in m)
+        """
+        # Convert meter to amps
+        value = numpy.multiply(value, self._amps_per_m)
+
+        # Subtract offset (center amplitude)
+        value = numpy.subtract(value, self._zero_offset_amps)
+
+        # Convert amps to 16-bit int
+        value = numpy.subtract(value, RANGE_AMPS[0])
+        value = int(numpy.multiply(value, 0xFFFF / (RANGE_AMPS[1] - RANGE_AMPS[0])))
+        # Note: in the example script a constant of 0.5 is added before converting
+        # to int to always round up (why?)
+
+        payload = [-value[0], -value[1], value[0], value[1]]
+
+        self._write_registers(payload)
```

Comment:
A setter *must* return the accepted value. Here that's useful here as it's not the original value requested, but the value rounded to an int (and converted back to meters).